### PR TITLE
chore: create template from `next-edge-ssr-loader`

### PR DESCRIPTION
**Pull request by ForsakenHarmony**
### Description

- Splits the generated code of the `next-edge-ssr-loader` out into 2 templates
- This PR also adds support for optional imports in templates

Closes WEB-1761

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: b773e652782d10df84733d83e1892cffd6b189eb

`pnpm test-dev test/development/app-hmr/hmr.test.ts` 
- app-dir-hmr > filesystem changes > should update server components pages when env files is changed (edge)

<details>
<summary>Expand output</summary>

  ● app-dir-hmr › filesystem changes › should update server components pages when env files is changed (edge)

    expect(received).toBe(expected) // Object.is equality

    Expected: "mac"
    Received: ""

      76 |         const envContent = await next.readFile(envFile)
      77 |         const browser = await next.browser('/env/edge')
    > 78 |         expect(await browser.elementByCss('p').text()).toBe('mac')
         |                                                        ^
      79 |         await next.patchFile(envFile, 'MY_DEVICE="ipad"')
      80 |
      81 |         try {

      at Object.<anonymous> (development/app-hmr/hmr.test.ts:78:56)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ForsakenHarmony** [comment]
Current dependencies on/for this PR:
* canary
  * **PR #56715** <a href="https://app.graphite.dev/github/pr/vercel/next.js/56715" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/vercel/next.js/56715?utm_source=stack-comment).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js hrmny/template-edge-ssr-loader | Change |
| - | - | - | - |
| buildDuration | 9.9s | 10s | N/A |
| buildDurationCached | 5.9s | 6s | N/A |
| nodeModulesSize | 172 MB | 172 MB | ⚠️ +48.2 kB |
| nextStartRea..uration (ms) | 536ms | 537ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js hrmny/template-edge-ssr-loader | Change |
| - | - | - | - |
| 199-HASH.js gzip | 27.5 kB | 27.5 kB | N/A |
| 3f784ff6-HASH.js gzip | 50.9 kB | 50.9 kB | ✓ |
| 99.HASH.js gzip | 182 B | 182 B | ✓ |
| framework-HASH.js gzip | 45.3 kB | 45.3 kB | ✓ |
| main-app-HASH.js gzip | 253 B | 251 B | N/A |
| main-HASH.js gzip | 32.9 kB | 32.9 kB | N/A |
| webpack-HASH.js gzip | 1.75 kB | 1.75 kB | N/A |
| Overall change | 96.4 kB | 96.4 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js hrmny/template-edge-ssr-loader | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js hrmny/template-edge-ssr-loader | Change |
| - | - | - | - |
| _app-HASH.js gzip | 206 B | 205 B | N/A |
| _error-HASH.js gzip | 182 B | 180 B | N/A |
| amp-HASH.js gzip | 506 B | 505 B | N/A |
| css-HASH.js gzip | 322 B | 323 B | N/A |
| dynamic-HASH.js gzip | 2.57 kB | 2.57 kB | N/A |
| edge-ssr-HASH.js gzip | 260 B | 259 B | N/A |
| head-HASH.js gzip | 350 B | 350 B | ✓ |
| hooks-HASH.js gzip | 369 B | 369 B | ✓ |
| image-HASH.js gzip | 4.35 kB | 4.35 kB | N/A |
| index-HASH.js gzip | 256 B | 256 B | ✓ |
| link-HASH.js gzip | 2.64 kB | 2.63 kB | N/A |
| routerDirect..HASH.js gzip | 312 B | 311 B | N/A |
| script-HASH.js gzip | 385 B | 384 B | N/A |
| withRouter-HASH.js gzip | 307 B | 308 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 1.08 kB | 1.08 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js hrmny/template-edge-ssr-loader | Change |
| - | - | - | - |
| _buildManifest.js gzip | 485 B | 482 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js hrmny/template-edge-ssr-loader | Change |
| - | - | - | - |
| index.html gzip | 528 B | 529 B | N/A |
| link.html gzip | 541 B | 542 B | N/A |
| withRouter.html gzip | 524 B | 524 B | ✓ |
| Overall change | 524 B | 524 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js hrmny/template-edge-ssr-loader | Change |
| - | - | - | - |
| edge-ssr.js gzip | 93.8 kB | 93.8 kB | N/A |
| page.js gzip | 152 kB | 152 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js hrmny/template-edge-ssr-loader | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 623 B | 623 B | ✓ |
| middleware-r..fest.js gzip | 150 B | 151 B | N/A |
| middleware.js gzip | 22.9 kB | 22.9 kB | N/A |
| edge-runtime..pack.js gzip | 1.92 kB | 1.92 kB | ✓ |
| Overall change | 2.55 kB | 2.55 kB | ✓ |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>page.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>edge-ssr.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 4dbfcd9e419febccda9b4c596809ed3400810bf2

**Review by sumit639-code — APPROVED** [review]

File: packages/next/src/build/entries.ts [hunk 1]
```diff
@@ -422,7 +422,7 @@ export function getEdgeServerEntry(opts: {
   }
 
   return {
-    import: `next-edge-ssr-loader?${stringify(loaderParams)}!`,
+    import: `next-edge-ssr-loader?${JSON.stringify(loaderParams)}!`,
```
Comment (timneutkens) [inline packages/next/src/build/entries.ts:425]: These needs to be querystring stringified as the parsing doesn't use JSON.

**Review by sokra — APPROVED** [review]

**Review by jridgewell — APPROVED** [review]

File: packages/next/src/build/entries.ts (hunk 1 above)
Comment (sokra) [reply packages/next/src/build/entries.ts:425]: @timneutkens might affect how webpack-internal references for this loader look like

**Pull request merged**